src/runtime-rs/crates/hypervisor/src/device/driver/virtio_net.rs

```diff
@@ -20,20 +28,106 @@ impl fmt::Debug for Address {
     }
 }
 
-#[derive(Debug, Clone)]
+#[derive(Clone, Debug, Default)]
 pub struct NetworkConfig {
+    pub index: u64,
     /// Host level path for the guest network interface.
     pub host_dev_name: String,
 
+    /// Guest iface name for the guest network interface.
+    pub virt_iface_name: String,
+
     /// Guest MAC address.
     pub guest_mac: Option<Address>,
 }
 
-#[derive(Debug, Clone)]
+#[derive(Clone, Debug, Default)]
 pub struct NetworkDevice {
     /// Unique identifier of the device
-    pub id: String,
-
+    pub device_id: String,
+    pub attach_count: u64,
     /// Network Device config info
     pub config: NetworkConfig,
 }
+
+impl NetworkDevice {
+    // new creates a NetworkDevice
+    pub fn new(device_id: String, config: &NetworkConfig) -> Self {
+        Self {
+            device_id,
+            attach_count: 0,
+            config: config.clone(),
+        }
+    }
+}
+
+#[async_trait]
+impl Device for NetworkDevice {
+    async fn attach(&mut self, h: &dyn hypervisor) -> Result<()> {
+        if self
+            .increase_attach_count()
+            .await
+            .context("failed to increase attach count")?
+        {
+            return Ok(());
+        }
+
+        if let Err(e) = h.add_device(DeviceType::Network(self.clone())).await {
+            self.decrease_attach_count().await?;
+            return Err(e);
+        }
+
+        return Ok(());
+    }
+
+    async fn detach(&mut self, h: &dyn hypervisor) -> Result<Option<u64>> {
+        if self
+            .decrease_attach_count()
+            .await
+            .context("failed to decrease attach count")?
+        {
+            return Ok(None);
+        }
+
+        if let Err(e) = h.remove_device(DeviceType::Network(self.clone())).await {
+            self.increase_attach_count().await?;
+            return Err(e);
+        }
+
+        Ok(Some(self.config.index))
+    }
+
+    async fn get_device_info(&self) -> DeviceType {
+        DeviceType::Network(self.clone())
+    }
+
+    async fn increase_attach_count(&mut self) -> Result<bool> {
+        match self.attach_count {
+            0 => {
+                // do real attach
+                self.attach_count += 1;
+                Ok(false)
+            }
+            std::u64::MAX => Err(anyhow!("device was attached too many times")),
+            _ => {
+                self.attach_count += 1;
+                Ok(true)
+            }
+        }
+    }
+
+    async fn decrease_attach_count(&mut self) -> Result<bool> {
```

Comment:
in my opinion, returning a bool is very confusing, why don't we just return the attach_count and leave the conditional logic to callers?

Reply:
As a result of so many similar increase/decrease operations, I think it's better to do so as it is.
And the `Result<bool>`  seems not so good, but the refactoring work is coming soon.